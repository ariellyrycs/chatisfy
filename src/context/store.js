import React, { useReducer, createContext, useContext } from 'react';
import Mock from '../Mock';
import {generateUId} from '../Utils/General.js';


const currentUser = {
	id: generateUId(),
	nickName: ''
};

const Context = createContext();
const useAppContext = () => useContext(Context);

const initialState = {
	...Mock,
	currentUser
};

const mainReduce = (state, action) => {
	switch (action.type) {

		case 'UPDATE_NICKNAME':
			return {
				...state,
				currentUser: {
					...state.currentUser,
					nickName: action.nickName
				}
			};
		default:
			return state;
	}
};

function ContextProvider(props) {
	const [state, dispatch] = useReducer(mainReduce, initialState);
	return (
		<Context.Provider value={{ state, dispatch }}>
			{props.children}
		</Context.Provider>
	);
};

export { ContextProvider, Context, useAppContext };