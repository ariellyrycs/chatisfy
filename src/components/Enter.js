import React, { useState } from 'react';
import { useAppContext } from '../context/store';
import { Button, Grid, Paper, FormControl, FormHelperText, InputLabel, Input } from '@material-ui/core';
import { makeStyles } from '@material-ui/core/styles';


const useStyles = makeStyles({
    spacing: {
        'margin-top': '20px',
        'margin-bottom': '20px'
    },
    fullHeight: {
        height: '100vh'
    }
});

function Enter({ setLog }) {
    const [nickName, setNickName] = useState('');

    const [error, setError] = useState('');

    const { state, dispatch } = useAppContext();

    const classes = useStyles();

    const procceed = (e) => {
        e.preventDefault();

        if (nickName.trim() === '') {
            setError('Should enter a valid nickName');
        } else {
            setLog(true);
            dispatch({ type: 'UPDATE_NICKNAME', nickName, sentBy: state.currentUser.id });
        }
        return false;
    };

    const handleEnter = e => {
        if (e.keyCode === 13 && !e.shiftKey) {
            procceed(e);
        }
        return false;
    };

    return (
        <Grid container justify='center' alignItems='center' className={classes.fullHeight}>
            <Grid item sm={6} md={3}>
                <Paper>
                    <form onSubmit={procceed}>
                        <Grid container justify='center' alignItems='center'>
                            <Grid item xs={10}>
                                <FormControl className={classes.spacing} fullWidth={true}>
                                    <InputLabel htmlFor='nickname-enter'>Enter your NickName</InputLabel>
                                    <Input
                                        id='nickname-enter'
                                        aria-describedby='Enter nickname'
                                        value={nickName}
                                        onChange={e => setNickName(e.target.value)}
                                        onKeyDown={handleEnter} />

                                    <FormHelperText
                                        id='nickname-enter-helper-text'
                                        error={true}
                                    >{error}
                                    </FormHelperText>
                                </FormControl>
                            </Grid>
                            <Grid item xs={10}>
                                <Button className={classes.spacing}
                                    id='enter-log'
                                    type='submit'
                                    variant='contained'
                                    color='primary'
                                    fullWidth={true}>
                                    Enter
                                </Button>
                            </Grid>
                        </Grid>
                    </form>
                </Paper>
            </Grid>
        </Grid>
    )
}
export default Enter;