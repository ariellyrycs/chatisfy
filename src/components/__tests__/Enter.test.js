import React from 'react';
import * as Store from '../../context/store';
import Enter from '../Enter';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json'
import { arrowFunctionExpression } from '@babel/types';


describe('<Enter />', () => {

    test('should render Enter', () => {
        const contextValues = {
            state: {
            },
            dispatch: () => {}
        };
        jest.spyOn(Store, 'useAppContext')
            .mockImplementation(() => contextValues);
        const tree = shallow(<Enter />);
        expect(tree.find('#nickname-enter-helper-text').text()).toBe('');
        tree.find('form').simulate('submit', {preventDefault: () => {}});
        expect(tree.find('#nickname-enter-helper-text').text()).toBe('Should enter a valid nickName');
    });

    test('should make the propert changes and trigget distach', () => {
        const dispatchStub = jest.fn();
        const setLogStub = jest.fn();
        
        const contextValues = {
            state: {
                currentUser: {
                    id: 2
                }
            },
            dispatch: dispatchStub
        };
        jest.spyOn(Store, 'useAppContext')
            .mockImplementation(() => contextValues);
        const tree = shallow(<Enter setLog={setLogStub}/>);
        expect(tree.find('#nickname-enter-helper-text').text()).toBe('');
        tree.find('#nickname-enter').simulate('change', { target: { value: 'Changed' } })
        tree.find('#nickname-enter').simulate('keydown', {keyCode: 13, shiftKey: false, preventDefault: () => {}});
        expect(dispatchStub).toHaveBeenCalledWith(
            expect.objectContaining({
                type: 'UPDATE_NICKNAME', nickName: 'Changed', sentBy: 2
            })
        );
        expect(setLogStub).toHaveBeenCalledWith(true);
    });

    test('should render chat room component', () => {
        const contextValues = {
            state: {
            },
            dispatch: () => {}
        };
        jest.spyOn(Store, 'useAppContext')
            .mockImplementation(() => contextValues);
        const tree = shallow(<Enter/>);
        expect(toJson(tree)).toMatchSnapshot();
    });
});