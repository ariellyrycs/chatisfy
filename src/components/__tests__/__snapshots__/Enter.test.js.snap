// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Enter /> should render chat room component 1`] = `
<WithStyles(ForwardRef(Grid))
  alignItems="center"
  className="makeStyles-fullHeight-2"
  container={true}
  justify="center"
>
  <WithStyles(ForwardRef(Grid))
    item={true}
    md={3}
    sm={6}
  >
    <WithStyles(ForwardRef(Paper))>
      <form
        onSubmit={[Function]}
      >
        <WithStyles(ForwardRef(Grid))
          alignItems="center"
          container={true}
          justify="center"
        >
          <WithStyles(ForwardRef(Grid))
            item={true}
            xs={10}
          >
            <WithStyles(ForwardRef(FormControl))
              className="makeStyles-spacing-1"
              fullWidth={true}
            >
              <WithStyles(ForwardRef(InputLabel))
                htmlFor="nickname-enter"
              >
                Enter your NickName
              </WithStyles(ForwardRef(InputLabel))>
              <WithStyles(ForwardRef(Input))
                aria-describedby="Enter nickname"
                id="nickname-enter"
                onChange={[Function]}
                onKeyDown={[Function]}
                value=""
              />
              <WithStyles(ForwardRef(FormHelperText))
                error={true}
                id="nickname-enter-helper-text"
              />
            </WithStyles(ForwardRef(FormControl))>
          </WithStyles(ForwardRef(Grid))>
          <WithStyles(ForwardRef(Grid))
            item={true}
            xs={10}
          >
            <WithStyles(ForwardRef(Button))
              className="makeStyles-spacing-1"
              color="primary"
              fullWidth={true}
              id="enter-log"
              type="submit"
              variant="contained"
            >
              Enter
            </WithStyles(ForwardRef(Button))>
          </WithStyles(ForwardRef(Grid))>
        </WithStyles(ForwardRef(Grid))>
      </form>
    </WithStyles(ForwardRef(Paper))>
  </WithStyles(ForwardRef(Grid))>
</WithStyles(ForwardRef(Grid))>
`;
